extensions/controllers/sandboxwarmpool_controller.go

```diff
@@ -0,0 +1,244 @@
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	extensionsv1alpha1 "sigs.k8s.io/agent-sandbox/extensions/api/v1alpha1"
+)
+
+const (
+	poolLabel = "agents.x-k8s.io/pool"
+)
+
+// SandboxWarmPoolReconciler reconciles a SandboxWarmPool object
+type SandboxWarmPoolReconciler struct {
+	client.Client
+}
+
+//+kubebuilder:rbac:groups=extensions.agents.x-k8s.io,resources=sandboxwarmpools,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=extensions.agents.x-k8s.io,resources=sandboxwarmpools/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile implements the reconciliation loop for SandboxWarmPool
+func (r *SandboxWarmPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// Fetch the SandboxWarmPool instance
+	warmPool := &extensionsv1alpha1.SandboxWarmPool{}
+	if err := r.Get(ctx, req.NamespacedName, warmPool); err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.Info("SandboxWarmPool resource not found. Ignoring since object must be deleted")
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Failed to get SandboxWarmPool")
+		return ctrl.Result{}, err
+	}
+
+	// Handle deletion
+	if !warmPool.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info("SandboxWarmPool is being deleted")
+		return ctrl.Result{}, nil
+	}
+
+	// Save old status for comparison
+	oldStatus := warmPool.Status.DeepCopy()
+
+	// Reconcile the pool (create or delete Pods as needed)
+	if err := r.reconcilePool(ctx, warmPool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Update status if it has changed
+	if err := r.updateStatus(ctx, oldStatus, warmPool); err != nil {
+		log.Error(err, "Failed to update SandboxWarmPool status")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// reconcilePool ensures the correct number of pods exist in the pool
+func (r *SandboxWarmPoolReconciler) reconcilePool(ctx context.Context, warmPool *extensionsv1alpha1.SandboxWarmPool) error {
+	log := log.FromContext(ctx)
+
+	// Get the pool label value (pool-name-hash)
+	poolLabelValue := warmPool.Name
+
+	// List all pods with the pool label
+	podList := &corev1.PodList{}
+	labelSelector := labels.SelectorFromSet(labels.Set{
+		poolLabel: poolLabelValue,
+	})
+
+	if err := r.List(ctx, podList, &client.ListOptions{
+		LabelSelector: labelSelector,
+		Namespace:     warmPool.Namespace,
+	}); err != nil {
+		log.Error(err, "Failed to list pods")
+		return err
+	}
+
+	desiredReplicas := warmPool.Spec.Replicas
+	currentReplicas := int32(len(podList.Items))
```

Comment:
Handle the case when podList contains pods whose `controllerRef` aren't set to the warmpool. If its controllerRef isn't set, adopt it; otherwise:

1. It shouldn't be counted in the status 
2. It's not part of the pool when determining pods to be created/deleted. 
3. Consider generating a warning log